File: android/lca/src/main/java/com/dreampany/lca/data/source/room/RoomNewsDataSource.java
/*
package com.dreampany.lca.data.source.room;

import com.dreampany.frame.misc.exception.EmptyException;
import com.dreampany.lca.data.misc.NewsMapper;
import com.dreampany.lca.data.model.News;
import com.dreampany.lca.data.source.api.NewsDataSource;
import com.dreampany.lca.data.source.dao.NewsDao;

import java.util.List;

import javax.inject.Singleton;

import io.reactivex.Maybe;

*/
/**
 * Created by Hawladar Roman on 7/5/2018.
 * BJIT Group
 * hawladar.roman@bjitgroup.com
 *//*


@Singleton
public class RoomNewsDataSource implements NewsDataSource {

    private final NewsMapper mapper;
    private final NewsDao dao;

    public RoomNewsDataSource(NewsMapper mapper,
                              NewsDao dao) {
        this.mapper = mapper;
        this.dao = dao;
    }

    @Override
    public boolean isEmpty() {
        return getCount() == 0;
    }

    @Override
    public Maybe<Boolean> isEmptyRx() {
        return Maybe.fromCallable(this::isEmpty);
    }

    @Override
    public int getCount() {
        return dao.getCount();
    }

    @Override
    public Maybe<Integer> getCountRx() {
        return null;
    }

    @Override
    public boolean isExists(News news) {
        return false;
    }

    @Override
    public Maybe<Boolean> isExistsRx(News news) {
        return null;
    }

    @Override
    public long putItem(News news) {
        return dao.insertOrReplace(news);
    }

    @Override
    public Maybe<Long> putItemRx(News news) {
        return Maybe.create(emitter -> {
            long result = putItem(news);
            if (emitter.isDisposed()) {
                return;
            }
            if (result == -1L) {
                emitter.onError(new EmptyException());
            } else {
                emitter.onSuccess(result);
            }
        });
    }

    @Override
    public List<Long> putItems(List<News> news) {
        return dao.insertOrReplace(news);
    }

    @Override
    public Maybe<List<Long>> putItemsRx(List<News> news) {
        return Maybe.fromCallable(() -> putItems(news));
    }

    @Override
    public int delete(News news) {
        return 0;
    }

    @Override
    public Maybe<Integer> deleteRx(News news) {
        return null;
    }

    @Override
    public List<Long> delete(List<News> news) {
        return null;
    }

    @Override
    public Maybe<List<Long>> deleteRx(List<News> news) {
        return null;
    }

    @Override
    public News getItem(String id) {
        return null;
    }

    @Override
    public Maybe<News> getItemRx(String id) {
        return null;
    }

    @Override
    public List<News> getItems() {
        return null;
    }

    @Override
    public Maybe<List<News>> getItemsRx() {
        return null;
    }

    @Override
    public List<News> getItems(int limit) {
        return dao.getItems(limit);
    }

    @Override
    public Maybe<List<News>> getItemsRx(int limit) {
        return dao.getItemsRx(limit);
    }
}
*/

File: android/lca/src/main/kotlin/com/dreampany/lca/data/source/room/RoomNewsDataSource.kt
package com.dreampany.lca.data.source.room

import com.dreampany.frame.misc.exception.EmptyException
import com.dreampany.lca.data.misc.NewsMapper
import com.dreampany.lca.data.misc.PriceMapper
import com.dreampany.lca.data.model.News
import com.dreampany.lca.data.source.api.NewsDataSource
import com.dreampany.lca.data.source.dao.NewsDao
import com.dreampany.lca.data.source.dao.PriceDao
import io.reactivex.Maybe
import java.util.concurrent.Callable

/**
 * Created by roman on 2019-07-14
 * Copyright (c) 2019 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
class RoomNewsDataSource constructor(
    val mapper: NewsMapper,
    val dao: NewsDao
) : NewsDataSource {

    override fun isEmpty(): Boolean {
        return getCount() == 0
    }

    override fun isEmptyRx(): Maybe<Boolean> {
        return Maybe.fromCallable({ this.isEmpty() })
    }

    override fun getCount(): Int {
        return dao.count
    }

    override fun getCountRx(): Maybe<Int> {
        return Maybe.empty()
    }

    override fun isExists(news: News): Boolean {
        return false
    }

    override fun isExistsRx(news: News): Maybe<Boolean> {
        return Maybe.empty()
    }

    override fun putItem(news: News): Long {
        return dao.insertOrReplace(news)
    }

    override fun putItemRx(news: News): Maybe<Long> {
        return Maybe.create { emitter ->
            val result = putItem(news)
            if (emitter.isDisposed) {
                return@create
            }
            if (result == -1L) {
                emitter.onError(EmptyException())
            } else {
                emitter.onSuccess(result)
            }
        }
    }

    override fun putItems(news: List<News>): List<Long>? {
        return dao.insertOrReplace(news)
    }

    override fun putItemsRx(news: List<News>): Maybe<List<Long>> {
        return Maybe.fromCallable { putItems(news) }
    }

    override fun delete(news: News): Int {
        return 0
    }

    override fun deleteRx(news: News): Maybe<Int> {
        return Maybe.empty()
    }

    override fun delete(news: List<News>): List<Long>? {
        return null
    }

    override fun deleteRx(news: List<News>): Maybe<List<Long>> {
        return Maybe.empty()
    }

    override fun getItem(id: String): News? {
        return null
    }

    override fun getItemRx(id: String): Maybe<News> {
        return Maybe.empty()
    }

    override fun getItems(): List<News>? {
        return null
    }

    override fun getItemsRx(): Maybe<List<News>> {
        return Maybe.empty()
    }

    override fun getItems(limit: Int): List<News>? {
        return dao.getItems(limit)
    }

    override fun getItemsRx(limit: Int): Maybe<List<News>> {
        return dao.getItemsRx(limit)
    }
}